dask_kubernetes/core.py

```diff
@@ -393,6 +396,73 @@ def __init__(
         self.kwargs = kwargs
         super().__init__(**self.kwargs)
 
+    def _clean_pod_templates(self):
+        for pod_template, pod_type in [
+            (self.pod_template, "worker"),
+            (self.scheduler_pod_template, "scheduler"),
+        ]:
+            if not pod_template and dask.config.get(
+                "kubernetes.{}-template".format(pod_type), None
+            ):
+                d = dask.config.get("kubernetes.{}-template".format(pod_type))
+                d = dask.config.expand_environment_variables(d)
+                pod_template = make_pod_from_dict(d)
+
+            if not pod_template and dask.config.get(
+                "kubernetes.{}-template-path".format(pod_type), None
+            ):
+                import yaml
+
+                fn = dask.config.get("kubernetes.{}-template-path".format(pod_type))
+                fn = fn.format(**os.environ)
+                with open(fn) as f:
+                    d = yaml.safe_load(f)
+                d = dask.config.expand_environment_variables(d)
+                pod_template = make_pod_from_dict(d)
+
+            if pod_type == "worker":
+                self.pod_template = pod_template
+            else:
+                self.scheduler_pod_template = pod_template
+
+        if not self.pod_template:
+            msg = (
+                "Worker pod specification not provided. See KubeCluster "
+                "docstring for ways to specify workers"
+            )
+            raise ValueError(msg)
+
+        if not self.scheduler_pod_template:
+            self.scheduler_pod_template = copy.deepcopy(self.pod_template)
+            self.scheduler_pod_template.spec.containers[0].args = ["dask-scheduler"]
+
+        self.pod_template = clean_pod_template(self.pod_template, pod_type="worker")
+
+        self.scheduler_pod_template = clean_pod_template(
+            self.scheduler_pod_template, pod_type="scheduler"
+        )
+
+    def _fill_pod_templates(self):
+        for pod_template, pod_type in [
+            (self.pod_template, "worker"),
+            (self.scheduler_pod_template, "scheduler"),
+        ]:
+            # Default labels that can't be overwritten
+            pod_template.metadata.labels["dask.org/cluster-name"] = self._generate_name
+            pod_template.metadata.labels["dask.org/component"] = pod_type
+            pod_template.metadata.labels["user"] = escape(getpass.getuser())
+            pod_template.metadata.labels["app"] = "dask"
+            pod_template.metadata.namespace = self._namespace
+
+            if self.env:
+                pod_template.spec.containers[0].env.extend(
+                    [
+                        kubernetes.client.V1EnvVar(name=k, value=str(v))
+                        for k, v in self.env.items()
+                    ]
+                )
+            pod_template.metadata.generate_name = self._generate_name
```

Comment:
Again I would prefer to return the template rather than overwrite values.

Reply:
As far as I can tell this requires to `copy.deepcopy` the template. Is that okay with you ?

Reply:
Sure.